app/scripts/lib/tx-verification/tx-verification-middleware.ts

```diff
@@ -0,0 +1,111 @@
+import { hashMessage } from '@ethersproject/hash';
+import { verifyMessage } from '@ethersproject/wallet';
+import type { NetworkController } from '@metamask/network-controller';
+import { rpcErrors } from '@metamask/rpc-errors';
+import { Json, JsonRpcParams, hasProperty, isObject } from '@metamask/utils';
+import {
+  JsonRpcRequest,
+  JsonRpcResponse,
+  JsonRpcEngineEndCallback,
+  JsonRpcEngineNextCallback,
+} from 'json-rpc-engine';
+import {
+  SIG_LEN,
+  TRUSTED_BRIDGE_SIGNER,
+} from '../../../../shared/constants/bridge';
+import { FIRST_PARTY_CONTRACT_NAMES } from '../../../../shared/constants/first-party-contracts';
+
+export type BridgeTxParams = {
+  chainId?: `0x${string}`;
+  data: string;
+  from: string;
+  to: string;
+  value: string;
+};
+
+/**
+ * Creates a middleware function that verifies bridge transactions from the
+ * Portfolio.
+ *
+ * @param networkController - The network controller instance.
+ * @returns The middleware function.
+ */
+export function createTxVerificationMiddleware(
+  networkController: NetworkController,
+) {
+  return function txVerificationMiddleware(
+    req: JsonRpcRequest<JsonRpcParams>,
+    _res: JsonRpcResponse<Json>,
+    next: JsonRpcEngineNextCallback,
+    end: JsonRpcEngineEndCallback,
+  ) {
+    if (
+      req.method !== 'eth_sendTransaction' ||
+      !Array.isArray(req.params) ||
+      !isValidParams(req.params)
```

Comment:
We can skip this middleware if the method isn't a send transaction, but should we instead `end` if `isValidParams` fails since that would mean a bad extension could theoretically mutate a transaction request enough to fail these validations and skip the verification, but potentially still be processed by subsequent middleware and controllers?

Reply:
I wrote most of this validation. I think we should fall back to our regular transaction handling if we can't validate the transaction in this middleware.